Support more specification details

scottsymes opened the pull request:
- Support negative values: 
    - `print(try! formatter.dateComponents(from: "-P2D")!.day)  // -2`
- Throw an error when trying to parse fractional values 
    - `try formatter.dateCompnonents(from: "P1.5D") // ISO8601DurationFormatter.Error.fractionalValuesNotSupported`
- Only append String elements required, or a `.zero` String (`"PT0S"`) if all elements empty
- Complete `Formatter` overrides with `string(for:)`

ScottSymes-ANZ committed:
Support negatives

ScottSymes-ANZ committed:
Only append required string elements

ScottSymes-ANZ committed:
Throw an error for fractional values

ScottSymes-ANZ committed:
complete Formatter `string(for:)` override

kkla320 commented on Sources/ISO8601DurationFormatter/DateComponents+ISO8601Duration.swift:
```diff
@@ -28,15 +28,39 @@ extension DateComponents {
     - returns: A [String](https://developer.apple.com/documentation/swift/string) object [ISO 8601 duration](https://en.wikipedia.org/wiki/ISO_8601#Durations) format using the current instance of [DateComponents](https://developer.apple.com/documentation/foundation/datecomponents)
      */
     public func toISO8601Duration() -> String {
```
I think it would to introduce a parameter where the user of the library can change the way non existing values are handled. With a parameter like this. I think of users, that maybe depend on the output having 0 for non existing values.

kkla320 commented on Sources/ISO8601DurationFormatter/DateComponents+ISO8601Duration.swift:
```diff
@@ -28,15 +28,39 @@ extension DateComponents {
+
+}
+
+private extension String {
```
I would like to move the method, to a new file, rename the method and make the parameters more generic. 
Something like this:
```swift
mutating func appendIfNotZeroOrNil(_ value: Int?, suffix: String)
```

kkla320 commented on Sources/ISO8601DurationFormatter/ISO8601DurationFormatter.swift:
```diff
@@ -23,55 +38,77 @@ public class ISO8601DurationFormatter: Formatter {
         return nil
     }
     
     public override func getObjectValue(_ obj: AutoreleasingUnsafeMutablePointer<AnyObject?>?, for string: String, errorDescription error: AutoreleasingUnsafeMutablePointer<NSString?>?) -> Bool {
```
I think, with the new throwing, I would move the functionality to an extension on `String` itself. The formatter API would just become a proxy for this extension. This would be a solution, similar to `toISO8601Duration`.

Commit:
- Added a new parameter for toISO8601Duration for emitting zero or nil values
- Created a new init for DateComponents for creating DateComponents from a ISO8601Duration string
- ISO8601DurationFormatter is now a proxy for DateComponents.toISO8601Duration and DateComponents.init(iso8601DurationString: String)
- Added new error handling for ISO8601DurationFormatter. Instead of nil, the method dateComponents now throws an error explaining what went wrong.

Commit:
Added new method to formatter for creating a string from DateComponents
Added more documentation

kkla320 commented:
@scottsymes @ScottSymes-ANZ I made some changes to your PullRequest to make the API more streamlined. 
- I moved the creation of a DateComponents from a string to a DateComponents extension
- The ISO8601DurationFormatter is now a proxy for the two extension methods of DateComponents
- Creation of the DateComponents now does not return an optional value, but fails with an `DateComponents.ISO8601ConversionErrors`
- I added an optional parameter to the Formatter and the toISO8601Duration method for dropping zero or nil values

Please let me know if the changes I made in addition to your PR are okay for you and satisfy your needs :)
